### ISSUE TITLE
database connection isn't set to UTC

### ISSUE THREAD

**Issue by plup** [psycopg/psycopg2#1293]
I just add this error after an upgrade. It's probably right about my database not being in UTC but I don't see the problem.

```
venv/bin/pip install -U psycopg2
Collecting psycopg2
  Downloading psycopg2-2.9.tar.gz (379 kB)
     |████████████████████████████████| 379 kB 7.2 MB/s 
Building wheels for collected packages: psycopg2
  Building wheel for psycopg2 (setup.py) ... done
  Created wheel for psycopg2: filename=psycopg2-2.9-cp36-cp36m-linux_x86_64.whl size=473570 sha256=c01f7407f6cc56a729d8693eb62133f284c8612b0ae0ef109d8bfa01ba5441d6
  Stored in directory: /home/test2/.cache/pip/wheels/70/01/46/39007954b343d1d72463a755e6a74e6106b10748fe22e2058b
Successfully built psycopg2
Installing collected packages: psycopg2
  Attempting uninstall: psycopg2
    Found existing installation: psycopg2 2.8.6
    Uninstalling psycopg2-2.8.6:
      Successfully uninstalled psycopg2-2.8.6
Successfully installed psycopg2-2.9
```

**Comment by dvarrazzo** [comment]
Sorry, where is the error?

**Comment by coderanger** [comment]
Just spent a while debugging this, Psycopg 2.9 changed the value passed to `tzinfo_factory` from an int to a timedelta. Django 2.2 (possibly newer but I'm on 2.2) has a check for `offset == 0` and since `timedelta(0) != 0` it goes boom. Not sure what the path forward here is but that's the cause I think.

**Comment by dvarrazzo** [comment]
Uhm, I checked and I thought that the change wouldn't have affected django. Looking into it...

**Comment by coderanger** [comment]
Looks like Django removed the check in 3.1 when it added support for non-UTC DB timezones. Django 3.0 is no longer supported but 2.2 is for another year-ish so if not fixing this on this side we'll need a 2.2 patch release :)

**Comment by coderanger** [comment]
https://github.com/django/django/blob/2.2.24/django/db/backends/postgresql/utils.py is the code on the django side.

**Comment by coderanger** [comment]
Arguably this is outside the purview of Django LTS support so I think you could just say "not a conflict" and maybe Django notes the compat issue in a blog post or something, since the fix of pinning to 2.8.6 until getting to Django 3 is easy :D

**Comment by dvarrazzo** [comment]
I see. I can take a look. However I don't see exactly how to fix the problem.

The change was needed to support timezone with fractional minutes, otherwise timestamps in the past have an error of up to 30 sec. (#1272). The integer passed to `FixedOffsetTimezone` is a number of minutes so it didn't allow to use seconds, as the `datetime.timezone` module of Python 3.7 and following allows (of course `FixedOffsetTimezone` predates (`datetime.timezone`).

One option to fix the problem would have been to pass a number of seconds, but that would have been dangerous as it could have been misinterpreted. But because now the stdlib has a datetime object it seems right to offer the same interface.

If there is a way to fix it on our side (suggestions welcome) I'm happy to introduce a change. However if only Django 2.2 is affected I am tempted to suggest people to just stick to `psycopg2 >= 2.8, < 2.9` in their requirements, as semver suggest to do, and have the issue fixed in a Django 2.2.x release, which might accommodate for the new value. Opening them a ticket and sending a MR...

**Comment by dvarrazzo** [comment]
https://code.djangoproject.com/ticket/32856#ticket

**Comment by coderanger** [comment]
Sounds good to me. Just one of the hazards of being on old versions of stuff :)

**Comment by dvarrazzo** [comment]
> Sounds good to me. Just one of the hazards of being on old versions of stuff :)

More an hazard of not making use of semver, IMO. This is not a change that would have been pushed in a bugfix release: people who constrained the library to the major version shouldn't have found the problem. I bet the OP has just `psycopg2` in their requirements file.

**Comment by coderanger** [comment]
2.8 to 2.9 is a feature release according to semver, not a major. For semver to catch this it would have had to be 3.0 instead. My pin was `~=2.8`.

**Comment by dvarrazzo** [comment]
That sounds true, and this change seems indeed a semver violation, but relying on things not changing in versions which are more than a bugfix release is pretty fragile. I have had more breakages than I would like to remember assuming that nothing bad would have happened going from x.y to x.(y+1). If you don't need new features it's safer to specify `>= x.y, < x.(y+1)`. Even more safer is to use Poetry or something similar (because breakages do happen in patch versions too).

I am looking forward to release psycopg 3, but that's a major rewrite, not a change in what I supposed to be an internal interface of an object predating the python datetime module (I don't think it was even documented that the object took an int number of minutes as constructor parameter).

**Comment by dvarrazzo** [comment]
MR ready in https://github.com/django/django/pull/14530. Tests pass ok, see https://djangoci.com/job/pull-requests-bionic/database=postgres,label=bionic-pr,python=python3.8/12017/console

**Comment by MotasemAghbar** [comment]
Same issue with django 2.2.x, fixed by downgrade psycopg2 to 2.8.6.

**Comment by elliottb303** [comment]
I am using Django 3.0.5

> I just add this error after an upgrade. It's probably right about my database not being in UTC but I don't see the problem.
> 
> ```
> venv/bin/pip install -U psycopg2
> Collecting psycopg2
>   Downloading psycopg2-2.9.tar.gz (379 kB)
>      |████████████████████████████████| 379 kB 7.2 MB/s 
> Building wheels for collected packages: psycopg2
>   Building wheel for psycopg2 (setup.py) ... done
>   Created wheel for psycopg2: filename=psycopg2-2.9-cp36-cp36m-linux_x86_64.whl size=473570 sha256=c01f7407f6cc56a729d8693eb62133f284c8612b0ae0ef109d8bfa01ba5441d6
>   Stored in directory: /home/test2/.cache/pip/wheels/70/01/46/39007954b343d1d72463a755e6a74e6106b10748fe22e2058b
> Successfully built psycopg2
> Installing collected packages: psycopg2
>   Attempting uninstall: psycopg2
>     Found existing installation: psycopg2 2.8.6
>     Uninstalling psycopg2-2.8.6:
>       Successfully uninstalled psycopg2-2.8.6
> Successfully installed psycopg2-2.9
> ```

Same with Django 3.0.5. Oddly my RDS pg database is set to UTC. Trying upgrade to Django 3.2..  

Django traceback:
/usr/local/lib/python3.9/site-packages/django/db/backends/postgresql/utils.py in utc_tzinfo_factory
        raise AssertionError("database connection isn't set to UTC")

**Comment by elliottb303** [comment]
@plup Was on Django 3.0.5, and AWS RDS pg DB was set to UTC.. So a strange error. 

Bumping Django to 3.2 and using psyopgg2 2.9.1 is looking good so far. 

Django==3.2
psycopg2-binary==2.9.1

Django trace I was having for more info..
/usr/local/lib/python3.9/site-packages/django/db/backends/postgresql/utils.py in utc_tzinfo_factory
        raise AssertionError("database connection isn't set to UTC")

**Comment by samoccupy** [comment]
> @plup Was on Django 3.0.5, and AWS RDS pg DB was set to UTC.. So a strange error.
> 
> Bumping Django to 3.2 and using psyopgg2 2.9.1 is looking good so far.
> 
> Django==3.2
> psycopg2-binary==2.9.1
> 
> Django trace I was having for more info..
> /usr/local/lib/python3.9/site-packages/django/db/backends/postgresql/utils.py in utc_tzinfo_factory
> raise AssertionError("database connection isn't set to UTC")

how do i downgrade to pyscopg been searching the net for the whole day, thanks

**Comment by dvarrazzo** [comment]
The Django team has established that they will not fix Django 2.2 to be compatible with psycopg 2.9 (https://github.com/django/django/pull/14530).

If you are using django 2.2 you mush use `psycopg2 < 2.9` in your requirements file.

**Comment by samoccupy** [comment]
Hi Daniele So grateful for your reply !,
But since ive been writing django ive not seen or come in contact with any
requirement file ?

**Comment by dvarrazzo** [comment]
@samoccupy you must have installed psycopg somehow. If you used a command such as `pip install psycopg2-binary`, manually or in a script, you should use `pip install "psycopg2-binary<2.9"` instead (if in a shell, quotes are mandatory to avoid shell redirections).

**Comment by samoccupy** [comment]
Thanks Daniele,
My terminal served me this Requirement already satisfied:
psycopg2-binary<2.9,
and then 6, in utc_tzinfo_factory
    raise AssertionError("database connection isn't set to UTC")

**Comment by Zajozor** [comment]
Hi, 

#1200 basically requires an M1 Mac to use >=2.9 - any recommendations on how to deal with this problem in that case?
Any way to use Django 2 + psycopg2 + M1 Mac ?

Thanks! :)

**Comment by dvarrazzo** [comment]
@Zajozor you can install psycopg from source after setting the env vars requested.

**Comment by Zajozor** [comment]
@dvarrazzo Thanks for the quick reply! I know that, I've been just wondering whether there is some quick way to prevent that - we install the library during a docker build process (docker uses arm64 now pretty well), which is executed often (and building psycopg increases the the image build time by 80-100%).

I guess that until we upgrade to Django>=3, we'll have to settle with that..

One nice thing I found out is that when using `psycopg2-binary` it uses the prebuilt binary on supported architectures, and falls back to building on arm64, so at least we can keep the workflow fast for people not having arm64 yet.

**Comment by zyv** [comment]
@dvarrazzo do you happen know if there is something like the RPM tag `Conflicts` in pip and whether it would be feasible to add a conflicts tag for Django < 3.x / psycopg 2.9+?

I think it would be really great to be able to declare a conflict with Django < 3.x on psycopg side, otherwise literally everyone in the world using Django 2 will have to pin psycopg < 2.9 at their end... and remember to unpin it during the next big upgrade. We did this, and it's quite some work to do for all our Django-based microservices!

**Comment by dvarrazzo** [comment]
@zyv I am not aware of any conflict mechanism, and frankly if there was it should be the dependant library to declare a dependency incompatible, not the other way around.

As per https://github.com/django/django/pull/14530, the Django maintainers consider that it is easier for Django 2.2 users to pin the dependencies than for them to fix the issue, which is a one-line change. If you think otherwise please let the Django maintainers know.

Personally, I would have fixed the issue as I proposed, however I sort of agree that if you are pinning Django to 2.2 it is reasonable to pin psycopg2 too.

**Comment by zyv** [comment]
@dvarrazzo sure, you are right, in the RPM world this mechanism is mostly used exactly the other way around and is basically considered as the means of last resort to save users the trouble if other packagers are uncooperative for whatever reasons.

Unfortunately, Django developers have a rather special LTS policy and have been refusing similar trivial code or syntax fixes in the past... I'm afraid we'll just have to deal with it - thanks for all your efforts & offering a backport though, it was still worth a try to save users time.

**Comment by erayerdin** [comment]
You can change poetry dependencies to `psycopg2 = "~2.8.6"`.

**Comment by triole** [comment]
We know about this problem. RDMO version 1.6 will not use the psycopg library anymore because is uses django3. The issue will disappear with the update. I will change the docker setup accordingly. 

A discussion can be found here: https://github.com/rdmorganiser/rdmo/issues/351

**Comment by elkd** [comment]
> Same issue with django 2.2.x, fixed by downgrade psycopg2 to 2.8.6.
> 
> Django Error: "database connection isn't set to UTC"

Just to highlight the conclusion for this issue, if you're on Django >=2.2.x, <3, you SHOULD PIN to Psycopg2 == 2.8.6. 
Nah fix for ya!
